Allow using branded oauth buttons

sherifnada opened the pull request:
## What
Certain platforms require 🤡 the use of specific brandings with oauth buttons. This PR introduces the ability to customize oauth buttons depending on the platform and defaults to the current style if there are no custom branding rules. 

example: 
Github (not branded)
<img width="811" alt="Screen Shot 2021-11-22 at 3 58 03 PM" src="https://user-images.githubusercontent.com/6246757/142952942-c1222e70-4f72-4c53-a3b4-f48e37a87297.png">

Google (branded)
<img width="875" alt="Screen Shot 2021-11-22 at 3 58 14 PM" src="https://user-images.githubusercontent.com/6246757/142952950-8de52ffa-07c9-471f-8c6d-2d5b65f5b244.png">

## How
Creates `AuthButtonFactory.tsx` which knows about custom brandings required and returns the appropriate button depending on the selected connector.

## Recommended reading order
1. `AuthButtonFactory.tsx`
2. `AuthButton.tsx`
3. `destination.ts`

## 🚨 User Impact 🚨
The only impact should be that custom branded buttons are seen in the connection UI.

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButton.tsx:
```diff
@@ -2,7 +2,7 @@ import React from "react";
 import styled from "styled-components";
 
-import { Button } from "components";
+import getButton from "./AuthButtonFactory";
```
No need to use any sort of factories here. It is just another component that customizes buttons. You should change it some that fits better frontend React code.

sherifnada replied:
Is there an example I can follow of what you're referring to here? I'm not sure what the correct pattern looks like

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+import { Button } from "../../../../components/base/Button";
+import { FormattedMessage } from "react-intl";
```
Absolute lib imports should go first

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+import { Button } from "../../../../components/base/Button";
+import { FormattedMessage } from "react-intl";
+import { ConnectorDefinition } from "../../../../core/domain/connector";
```
all `../../../` could be removed.

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+  } else if (selectedService && isDestinationDefinition(selectedService)) {
+    definitionId = selectedService.destinationDefinitionId;
+  } else {
+    throw new Error(`Unrecognized service type. Service: ${selectedService}`);
```
Very bad idea to throw an error. ConnectorDefinition is always either SourceDefinition or DestinationDefinition: how is it possible to throw an error then?

sherifnada replied:
It generally should not happen. This is to future proof the case where let's say we add a new type of definition and it is not handled here for whatever reason. I can remove the error though if that's not a pattern followed here

Jamakase replied:
I think in such cases we should better display users something like an unexpected error happened or something like that. Because in case error is thrown this way - the app could fail ( or user will be redirected back to the last working state ). So instead of handling unexpected case - we will fail silently when an error is thrown and will have to somehow handle such case in `<ErrorBoundary>`

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  switch (definitionId) {
+    case "253487c0-2246-43ba-a21f-5116b20a2c50": // google ads
```
I think it's better to extract it to either a separate method or file and call it like `isGoogleOauth`, so if we add more google connectors - we will change that file only and won't change components implementation.

sherifnada replied:
Sounds good. I can remove the switch and do the suggested approach

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+
+  switch (definitionId) {
+    case "253487c0-2246-43ba-a21f-5116b20a2c50": // google ads
+    case "eff3616a-f9c3-11eb-9a03-0242ac130003": // google analytics
```
I believe that falling through in switch cases is actually a bad pattern and `[].includes()` is much easier to read.

sherifnada replied:
Will do

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+            <img
+              src="/connectors/google/btn_google_signin_dark_normal_web@2x.png"
+              height={40}
+              alt={"Sign in with google"}
```
We use i18n for all messages, so it should use formatMessage

sherifnada replied:
@Jamakase this is almost certainly me not knowing front end very well - the Google image contains the "authenticate with google" text as well as the logo. Are you saying I should instead use the logo only next to our text? In this case do you mind giving a quick code example of what that would look like? Do I still add an img tag inside the button? How would I style the button (text/background/font etc) in a custom way while using <Button />? Any quick code sample would be helpful

Jamakase replied:
I just meant that the alt text you provided doesn't support i18n. We use react-intl lib fo rit. So you need to add this text into en.json and then use id as `formatMessage("some.id")` instead of providing text explicitly

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+        <Button isLoading={loading} type="button" onClick={run} iconOnly>
+          {
+            <img
+              src="/connectors/google/btn_google_signin_dark_normal_web@2x.png"
```
When this is moved to the proper place, path will look much better.

sherifnada replied:
I put the assets in `public/connectors/google` -- are you saying that it should not be in public?

Jamakase replied:
oh, I thought it was put somewhere in src folder. Sorry for confusion.

Jamakase commented on airbyte-webapp/src/views/Connector/ServiceForm/components/AuthButtonFactory.tsx:
```diff
@@ -0,0 +1,58 @@
+          {
+            <img
+              src="/connectors/google/btn_google_signin_dark_normal_web@2x.png"
+              height={40}
```
I am not sure that we set styles somewhere on the node directly, have you seen that we set sizes this way somewhere in code? I think we try to use styled-components everywhere

sherifnada replied:
I saw it somewhere in the code (the example was displaying cactus.png). It was a helpful way to control the size of the button because otherwise the image was a little too big

Jamakase replied:
ok, let's then keep it as is.

Jamakase reviewed (changes requested):
First of all, I believe that you still use node version 14, that's why your package-lock was updated. You need to either upgrade your node version or do not commit this file.

Secondly, we try to keep all binaries out of our code ( images, fonts and etc ), so you need to move .png files to the proper place.

And the last, but not the least important: how are going to handle cases, when user added some custom connector and set all oauth required info? I believe we should display google oauth buttons in such case too. Isn't it better to handle such things in a less error-prone way? 
Like return such info from backend, that button is overridden? Or at least add to spec that some oauth is part of google oauth - then we won't need to add ids of specs directly. What you have done here looks like a very dirty hack.

sherifnada commented:
@Jamakase thanks for the feedback. A couple of notes:
1. Where should I put the png? I put these assets here because this is primarily a styling concern (we might need custom css in the future etc) so it would be overreaching/cumbersome for the spec/backend to send this information. 
2. I think we can safely disregard custom connectors. Really this is added to appease some very specific situations where platforms (cough google) which are not approving our oauth application until we've done this styling. Does that change how you view this at all?

Jamakase commented:
> @Jamakase thanks for the feedback. A couple of notes:
> 
> 1. Where should I put the png? I put these assets here because this is primarily a styling concern (we might need custom css in the future etc) so it would be overreaching/cumbersome for the spec/backend to send this information.
> 2. I think we can safely disregard custom connectors. Really this is added to appease some very specific situations where platforms (cough google) which are not approving our oauth application until we've done this styling. Does that change how you view this at all?

1. We put assets in public directory.
2. ok, so we don't need to take care of custom connectors? But anyway ideally we should better keep it somewhere on backend ( maybe part of spec ), so we won't need to change FE code everytime.

sherifnada commented:
@Jamakase I believe I've addressed all the feedback, except a couple of points: 

1. Where to put PNGs? They are currently in `airbyte-webapp/public/connectors/google` -- is this the wrong place? The path is somewhat ugly (`/connectors/google/<>.png`) but it seemed better than to put them at the root level
2. I don't anticipate that this will be done super frequently, but because there are style-heavy concerns (e.g: if we wanted to replicate `GoogleAuthButton.tsx` for other connectors) it seems like the frontend is the better place to put this? In either case, do you think we can move forward on the current impl and revisit putting this in the backend if it becomes an issue? 

I'm also having an issue with fonts. I added the Roboto-Medium font (pattern matched from existing fonts) but when I run the app locally I just can't get the font to display properly: 

![Screen Shot 2021-11-23 at 3 17 25 PM](https://user-images.githubusercontent.com/6246757/143145479-8badcc8b-34d1-40a5-a0f0-7debd405533d.png)

I set `font-family: "Roboto", sans-serif;` but in the screenshot below under the "rendered fonts" footer it seems to be rendering Helvetica. It is indeed different looking from the button on Google's style guidelines: 

![image](https://user-images.githubusercontent.com/6246757/143145565-d973599e-0779-4027-8838-203a9dac3dc1.png)

Do you have any advice on how to resolve this?

Jamakase reviewed (approved):
Awesome!

Jamakase commented:
> @Jamakase I believe I've addressed all the feedback, except a couple of points:
> 
> 1. Where to put PNGs? They are currently in `airbyte-webapp/public/connectors/google` -- is this the wrong place? The path is somewhat ugly (`/connectors/google/<>.png`) but it seemed better than to put them at the root level

Yeah, let's keep them there. I think it's better than root anyway.

> I'm also having an issue with fonts. I added the Roboto-Medium font (pattern matched from existing fonts) but when I run the app locally I just can't get the font to display properly:
> I set `font-family: "Roboto", sans-serif;` but in the screenshot below under the "rendered fonts" footer it seems to be rendering Helvetica. It is indeed different looking from the button on Google's style guidelines:
> Do you have any advice on how to resolve this?

Hmm, I will take a look. I remember that I had some problems with fonts too.

sherifnada commented:
@Jamakase any luck with the fonts issue?

Jamakase commented:
@sherifnada pushed a fix for index.css import. It was working not correct somehow.
